torch/distributed/distributed_c10d.py

```diff
@@ -1122,6 +1144,37 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def all_gather_object(object, group=group.WORLD, async_op=False):
+    """
+    Performs all_gather on a picklable object.
+    """
+    if _rank_not_in_group(group):
+        return
+    _check_default_pg()
+
+    byte_tensor, size = _object_to_tensor(object)
+
+    group_size = torch.distributed.get_world_size(group=group)
+    object_size_list = [torch.LongTensor([0]) for _ in range(group_size)]
+
+    # collect the object sizes from all the tensors
+    all_gather(object_size_list, size, group=group)
+
+    # create the list of output tensors needed for all_gather
+    max_object_size = max(object_size_list)
+    tensor_list = [torch.ByteTensor(size=(max_object_size,)) for _ in range(group_size)]
+
+    if size != max_object_size:
+        # all tensors input into gather should be of the same size, so pad the local tensor if it is not the max size.
+        padding = torch.ByteTensor(size=(max_object_size - size,))
+        byte_tensor = torch.cat((byte_tensor, padding), dim=0)
+
+    all_gather(tensor_list, byte_tensor, group=group, async_op=async_op)
+    # unpickle the tensors back into objects.
+    objects = [_tensor_to_object(tensor, object_size) for object_size, tensor in zip(object_size_list, tensor_list)]
```

Comment:
What if this is an async_op? data in tensor_list might not be ready yet.

Reply:
Let's not support async operations for the "object collectives" to begin with.